### PR TITLE
fix(#224): address issue with expression tokenization

### LINKED ISSUE #224: 🐛 BUG: Astro compiler fails after first curly brace closure in an expression
### What version of `astro` are you using?

0.21.13

### What package manager are you using?

npm

### What operating system are you using?

Mac

### Describe the Bug

The following Astro file gets extra `$$render` functions added to the second and third functions within the expression.

```astro
{
	() => {
		let generate = (input) => {
			let a = () => { return; };
			let b = () => { return; };
			let c = () => { return; };
		};
	}
}
```

It gets turned into something like:

```astro
{
	() => {
		let generate = (input) => {
			let a = () => { return; };
			let b = () => $$render`${ return; }`;
			let c = () => $$render`${ return; }`;
		};
	}
}
```

Credit to @okikio for discovering the issue. 🙏

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-gjpgyq

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

- Closes #224
- Fixes logical error in expression tokenization step. It worked in most situations but #224 was a reproducible edge case where it failed.

## Testing

Added tokenizer and printer tests to verify that the case in the issue works. Existing tests ensure this is not a regression.

## Docs

N/A, bug fix only.

### CODE REVIEW

File: internal/token.go [hunk 1]
```diff
@@ -1674,9 +1674,9 @@ expression_loop:
 				z.tt = TextToken
 				return z.tt
 			}
-			z.openBraceIsExpressionStart = true
 			z.expressionStack[len(z.expressionStack)-1]--
 			if z.expressionStack[len(z.expressionStack)-1] == -1 {
+				z.openBraceIsExpressionStart = true
```
Comment (natemoo-re) [inline internal/token.go:1679]: Pretty simple fix! This edge case exposed that `z.openBraceIsExpressionStart = true` should only be reset when we've exhausted the entire `z.expressionStack`

File: internal/token_test.go [hunk 2]
```diff
@@ -122,6 +122,17 @@ func TestBasic(t *testing.T) {
...
+		{
+			"expression with multiple returns",
+			`<div>{() => {
+	let generate = (input) => {
+		let a = () => { return; };
+		let b = () => { return; };
+		let c = () => { return; };
+	};
+}}</div>`,
+			[]TokenType{StartTagToken, StartExpressionToken, TextToken, TextToken, TextToken, TextToken, TextToken, TextToken, TextToken, TextToken, TextToken, TextToken, TextToken, TextToken, TextToken, TextToken, TextToken, TextToken, EndExpressionToken, EndTagToken},
```
Comment (natemoo-re) [inline internal/token_test.go:134]: This was previously being tokenized incorrectly.

File: internal/printer/printer_test.go [hunk 3]
```diff
@@ -406,6 +406,34 @@ $$render` + "`" + `<div${$$addAttribute(color, "id")}>color</div>` + "`" + `
...
+}
+</div>`,
+			want: want{
+				code: `<html><head></head><body><div>
+${
+	() => {
+		let generate = (input) => {
+			let a = () => { return; };
+			let b = () => { return; };
+			let c = () => { return; };
```
Comment (natemoo-re) [inline internal/printer/printer_test.go:433]: Added a printer test as a sanity check, too.

**Review by jonathantneal — APPROVED** [review]
Thank you! The fix was so small that I had to look twice!
